create.sh

```diff
@@ -314,6 +314,30 @@ else
     default_series="$series"
 fi
 
+apt_proxy=
+snap_http_proxy=
+snap_https_proxy=
+if [[ -n ${http_proxy} ]]; then
+    if echo $http_proxy | grep -q :; then
```

Comment:
This does not address the case where `http_proxy` is set to `http://some.address[:port]`, i.e. where that string contains 2 `:`. On the other hand, now that I am trying to find if including a protocol is even allowed I can't find anything on the subject. Maybe I have been using `http_proxy` wrong this whole time :smile:

Reply:
I think I was assuming that the http_proxy variable was always set to a pure address since this code was what I originally found. In a previous PR I generalized PROXY_ADDRESS to be settable but still assumed it was an address.

```
maas-test-setup-new.sh:PROXY_ADDRESS=squid-deb-proxy.virtual
maas-test-setup-new.sh:PROXY=http://${PROXY_ADDRESS}:8080
```

That being said, I did not dig too deep into snap_http_proxy and what they expected, but the fact that there is both a snap_http_proxy and a snap_https_proxy might have made me think the http:// or https:// part of the URL was added somewhere else and all you pere proving was an IP or hostname.

Reply:
Looks like the snap stuff does accept a full URL. https://askubuntu.com/questions/764610/how-to-install-snap-packages-behind-web-proxy

Hmmm.... I suppose I can make the parsing sophisticated enough to parse any of:

IP_address
DNS_name
IP_address:port_number
DNS_name:port_number
http://IP_address
http://DNS_name
http://IP_address:port_number
http://DNS_name:port_number

A lot of https_proxy I see is uses http:// for the proxy URL. Do you know if there is ever a proxy URL that is https:// (where I assume the HTTP client is smart enough to require the TLS cert to match the proxy name???). In other words, would I need to be able to parse:

https://IP_address
https://DNS_name
https://IP_address:port_number
https://DNS_name:port_number

Reply:
(I am currently explicitly not wanting to parse port_name, i.e. looking up stuff in /etc/services)

Reply:
I found this post: <https://gist.github.com/joshisa/297b0bc1ec0dcdda0d1625029711fa24>. Looks somewhat complicated because there could be other fields in the URI such as a username for example.

Reply:
You stated in the initial PR description that the reason for this change is so that you can use a different port for the proxy. Can you specify the port as part of the `http_proxy` value? For example:
```console
http_proxy=http://my.synology.proxy:3128`
```

Reply:
What I am trying to ask is whether it is necessary for the script to be able to split off the port or set a default for the port in case `http_proxy` does not contain one already.

Reply:
Absolutely. In fact, that's effectively what I was doing. I just incorrectly assumed the http:// was added later (which is was, in the maas-test-setup-new.sh). 

The bigger question is do we want to force the user to provide a fully-valid URL including port up to but not including the root path, or do you want to make the port optional and if so, what should the default port be?

Reply:
I am leaning towards letting the user worry about a correctly formatted proxy
URI. But you are right, that means that the proxy is not set correctly in
`maas-test-setup-new.sh`. It would be better if we removed the explicit port
there.

What do you think?

Reply:
Agreed. Push setting the "correct" URL as early as possible, which would be from the user directly, then just pass that around everywhere. That will actually simplify the code, too. 

I'll rework the PR.